OpenRA.Game/Graphics/TerrainSpriteLayer.cs

```diff
@@ -45,25 +65,58 @@ public TerrainSpriteLayer(World world, WorldRenderer wr, Sheet sheet, BlendMode
 			map = world.Map;
 			rowStride = 6 * map.MapSize.X;
 
-			vertices = new Vertex[rowStride * map.MapSize.Y];
-			vertexBuffer = Game.Renderer.Device.CreateVertexBuffer(vertices.Length);
+			// We can use a buffer to track updates. We can then only send updates in the visible region to the GPU.
+			// This saves CPU by only sending updated rows when they become visible. This requires keeping a full copy
+			// of the vertices, so it doubles the memory required.
+			this.buffered = buffered;
+
+			if (buffered)
+				dirtyRows = new HashSet<int>();
+
+			vertices = new Vertex[buffered ? rowStride * map.MapSize.Y : 6];
+			vertexBuffer = Game.Renderer.Device.CreateVertexBuffer(rowStride * map.MapSize.Y);
 			emptySprite = new Sprite(sheet, Rectangle.Empty, TextureChannel.Alpha);
 
 			wr.PaletteInvalidated += UpdatePaletteIndices;
 		}
 
 		void UpdatePaletteIndices()
```

Comment:
How common is this case?

Reply:
During regular gameplay, this is unreachable. All the palettes are loaded upfront and therefore this never fires.

`PaletteInvalidated` was added in #8125:
> fixes a bug that would cause the shroud and terrain palettes to break if too many players were added in the editor

The editor is currently the only thing that updates palette after initial loading, and therefore the only place where this can fire.